hera_pspec/pspecdata.py

```diff
@@ -775,74 +844,152 @@ def pspec(self, bls, beam=None, input_data_weight='identity', norm='I',
-        if self.primary_beam != None:
+        if self.primary_beam is not None:
             scalar = self.scalar(taper=taper, little_h=True)
+        else: raise_warning("Warning: self.primary_beam is not defined, so pspectra are not properly normalized", verbose=verbose)
 
-        pvs = []; pairs = []
-        # Loop over pairs of datasets
-        for m in xrange(len(self.dsets)):
-            for n in xrange(m+1, len(self.dsets)):
-                # Datasets should not be cross-correlated with themselves, and
-                # dataset pair (m, n) gives the same result as (n, m)
-
-                # Loop over baselines
-                for bl in bls:
-                    if isinstance(bl, list):
-                        key1 = [(m,) + _bl for _bl in bl]
-                        key2 = [(n,) + _bl for _bl in bl]
-                    else:
-                        key1 = (m,) + bl
-                        key2 = (n,) + bl
-                    if verbose: print("Baselines:", key1, key2)
-
-                    # Set covariance weighting scheme for input data
-                    if verbose: print("  Setting weight matrix for input data...")
-                    self.set_R(input_data_weight)
-
-                    # Build Fisher matrix
-                    if verbose: print("  Building G...")
-                    Gv = self.get_G(key1, key2, taper=taper)
-
-                    # Calculate unnormalized bandpowers
-                    if verbose: print("  Building q_hat...")
-                    qv = self.q_hat(key1, key2, taper=taper)
-
-                    # Normalize power spectrum estimate
-                    if verbose: print("  Normalizing power spectrum...")
-                    Mv, Wv = self.get_MW(Gv, mode=norm)
-                    pv = self.p_hat(Mv, qv)
-                    
-                    # Multiply by scalar
-                    if self.primary_beam != None:
-                        if verbose: print("  Computing and multiplying scalar...")
-                        pv *= scalar
-
-                    # Save power spectra and dataset/baseline pairs
-                    pvs.append(pv)
-                    pairs.append((key1, key2))
+        # ensure both bls1 and bls2 are the same type
+        if isinstance(bls1[0], tuple) and isinstance(bls1[0][0], (int, np.int)) \
+            and isinstance(bls2[0], tuple) and isinstance(bls2[0][0], (int, np.int)):
+            # bls1 and bls2 fed as list of bl tuples
+            fed_bl_group = False
+
+        elif isinstance(bls1[0], list) and isinstance(bls1[0][0], tuple) and isinstance(bls2[0], list) \
+            and isinstance(bls2[0][0], tuple):
+            # bls1 and bls2 fed as list of bl groups
+            fed_bl_group = True
+            assert len(bls1) == len(bls2), "if fed as list of bl groups, len(bls1) must equal len(bls2)"
+
+        else:
+            raise TypeError("bls1 and bls2 must both be fed as either a list of bl tuples, or a list of bl groups")
```

Comment:
Does this mean that we don't support mixed lists, like `bls1 = [[tuple, tuple], tuple, tuple]`?

Reply:
the answer is yes, `bls1` should be fed as either `[tuple, tuple, ..., tuple]` or `[[tuple, tuple, ...], [tuple, ...], ..., [tuple]]`, but what you describe above can be achieved by wrapping a lone-tuple within a list. its up to the user to know to do that (and also `redcal` will automatically do that for you anyways)